### ISSUE TITLE
The design of `RoutableDoc` for Bulk Action

### ISSUE THREAD

**Issue by Gillani0** [CrowdStrike/opensearchtools#29]
I have few comments for `RoutableDoc`. For me it seems a bit intrusive to implement this interface for the json data within a bulk request. If I've a model that is shared between different services (each is using for different internal usage storage, query, etc), implementing `RoutableDoc` exposes the internals of my repository to these services. The caller has to understand that `index` is something internal to OpenSearch and doesn't have any specific meaning in that model. I can create a generic model without `Index` variable and then use a wrapper struct on top of it to implement `RoutableDoc`. However, this is again an extra object for the sake of implementing `RoutableDoc`. Wouldn't it be simpler to have json.Raw or IoReader instead and provide the dedicated fields `index` and `id` within bulk action? Furthermore, there can be cases where the `Id` of the document can be derived from a different logic (`ID()` and `Index()` functions are called multiple times so a dynamic implementation of them is not optimal) or the Id generation is left to OpenSearch. WDYT?

**Comment by cs-marshall** [comment]
A couple different points to break down.

> Model that is shared between different services

I would argue this model should not implement `RoutableDoc`. Like you said, `RoutableDoc` is very much an implementation of needs for OpenSearch. For the majority of usecases, a document needs to be routable to a specific index. So that implementation should be left to the internal model that interfaces with OpenSearch.

Not a very strong argument, but this is same pattern that popular ElasticSearch libraries use like the [olivere elastic](https://github.com/olivere/elastic) library.

> Id generation is left to OpenSearch

This is definitely a bug we have in our flow. I don't think we can do a create bulk action without providing a non empty id. Which is not valid, since you can leave OpenSearch to generate the ID for a new document.

> dynamic implementations of `ID()` and `Index()`

I think if they're dynamic or simple accessors is up to the implementers of the interface.

**Comment by cs-marshall** [comment]
I think overall I'm not very attached to the `RoutableDoc` being a necessity for the bulk actions. I do think that it is convenient, but that convenience is built a lot from habit of my past experience with the olivere elastic library.

**Comment by Gillani0** [comment]
>For the majority of usecases, a document needs to be routable to a specific index.

I agree in majority of some cases this is true. Coming from my experience, there can be cases where a shared model is used by multiple services to keep things consistent. For this case, a wrapper can be used that implement `RoutableDoc`, but just from my point of view,  it could have been simpler to use raw json. 

>I think if they're dynamic or simple accessors is up to the implementers of the interface.
So what I meant what is if we have something like: 

```
func Index(){
// do some preprocessing 
// get the timestamp and generate an index on it
// etc, etc
}
```
then it will not be very efficient. And the caller should be somehow informed to avoid such implementation.

**Comment by cs-marshall** [comment]
@jeffmclean-cs , @dylan-bourque - any thoughts on if we should drop `RoutableDoc` from the bulk actions?

**Comment by cs-marshall** [comment]
As for the efficiency issue. I'm not sure that's a concern of the library or if we need additional text. But that could be my bias from writing it.

```
type RoutableDoc interface {
	// Index returns the index the document should be routed to
	Index() string
	// ID returns the document ID
	ID() string
}
```

To me, that doc comment indicates the methods should just be accessors, not generators or more complex logic. But if the consumer wants to do something more complicated, that's completely up to them. There probably is some optimization we could do to reduce the count of calls to both interface methods, but I'm not sure if it's that concerning.

**Comment by dylan-bourque** [comment]
I think removing `RoutableDoc` would be a bad direction.

The main reason is that switching to "raw" JSON converts a compile time error (you must pass in a type that can provide an ID and index) to a runtime one where we'll have to inspect the generated JSON for correctness.

> Id generation is left to OpenSearch

I assume that one would let the Open Search server generate IDs by not passing one.  With `RoutableDoc`, that's trivial to implement by having a `ID()` return an empty string.  If we have code that's erroring out for that case then we can re-evaluate that logic.

On the topic of efficiency, *something* will have to do the work of determining which index each document should be routed to.  This interface defines that contract as concisely as possible and the expectation is that consumers would implement that logic.

For both ID and index, the `DocumentRef` type is a trivial implementation but any consumer of the library is free (encouraged, in fact) to provide their own if there's more logic needed.

>  a model that is shared between different services

IMO this goes against the fundamental idea of microservices.  Two services should not be sharing a model.  I'd even go so far as to argue that the storage and API layers of a single service shouldn't be sharing model types.

**Comment by Gillani0** [comment]
>  we'll have to inspect the generated JSON for correctness.

With having ioreader to consume the json and index and id additional variables, then I don't think we'll have to inspect the generate Json: if the provided body of the json is not correct then that's on the caller. I agree having a struct implementing reduces this issues and body is always generated from the passed model. 

> I assume that one would let the Open Search server generate IDs by not passing one

For the use case, where all the Bulk actions are routed to the same index (provided in the request) and IDs will be generated by OpenSearch (used in a plenty of usecases). The dummy implementation of Index() and ID() will be useless. True we can have empty string for IDs but wouldn't it be cleaner to have an option in the request that says if the IDs generation is delegated to OpenSearch or not?

> For both ID and index, the DocumentRef type is a trivial implementation

True but by the first look, without going into the doc details, It doesn't really shows if these should be used only as accessor or not. I learnt by doing some tests that Index() is called multiple times, hence, I shouldn't not have any complex logic defined within Index(). For me `RoutableDoc` looks more like a Java interface where methods are generally the accessors :) 

> IMO this goes against the fundamental idea of microservices.

True, but the anti-pattern of sharing models between services (usually in monorepo) (at storage level -->  ingestor and api) is commonly used. Sharing has both pros and cons (both at services and storage, API level).

This is just from my PoV and I think it could have been simpler (cleaner) for the caller to have dedicated variables for Index, ID instead of using the interface implementation.

**Comment by cs-marshall** [comment]
Ultimately you want a 
```
NewCreateWithoutRoutingBulkAction(doc any)
```

Where it would no-op on both index and id? I don't think that's a terrible utility usecase to support.

> It doesn't really shows if these should be used only as accessor or not

If it's just an accessor or not is entirely up to the consumer. All the library cares about is getting the index and id value. I think I still fail to see the problem with this. 

> RoutableDoc looks more like a Java interface where methods are generally the accessors

It absolutely is. Otherwise we couldn't have the `RoutableDoc` anything. We'd have to either have three methods for everything that uses a document:
```
// names are bad here
New*WithDoc(doc any)
New*WithDocAndID(id string, doc any)
New*WithDocAndFullRouting(id, index string, doc any)
```

Or always ask for all three items. By having the "accessor" java-like interface, this allows us to have a simpler more generic signature.

**Comment by Gillani0** [comment]
Ok got it.
What I was referring to have something like this:

```
type BulkAction struct {
	Type BulkActionType
	Doc  IoReader
        Index string
        ID   string
}
```
and then

`NewCreateBulkAction(doc IoReader, index, ID string)` and. `NewCreateBulkAction(doc IoReader)`

Same for the reader part: `opensearchtools.ReadDocument` takes the Routable Doc by we don't need index in it. The search request should be based on the index or alias. 

But I agree moving away from Routable Doc will require quite some work.

**Comment by cs-marshall** [comment]
Having the `opensearchtools.ReadDocument` use RoutableDoc is the same reason why the write side uses readable doc. Consistency. If the consumer has already implemented the interface to write their document, then that gives us a solid type to unmarshal into. I can't imagine a consumer would have a different write document than read document.
